Simple support for third party packages.

ntoll opened the pull request:
This PR fixes #675.

It has been tested on Linux (Ubuntu 18.04) but not Windows, OSX or Raspberry Pi. **Please give feedback from testing in these environments**.

# What does it do?

This is the simplest possible third party package manager for a beginners' editor. Until now we have bundled a bunch of "third party" modules with Mu in the official installers, thus bloating Mu's size. Furthermore, our bundled modules go out of date and/or miss some important packages for folks. This new functionality allows users to simply specify packages to be installed from PyPI. It also allows users to remove packages.

It is accessed via the "admin" dialog (accessed via the cog on the bottom right of the UI) and is one of the tabs found therein. The following GIF demonstrates it in action:

![mu_modules](https://user-images.githubusercontent.com/37602/51917850-5c4d7900-23d8-11e9-8ba6-16800c9308f0.gif)

The user simply ensures the list of module names (one per line) reflects their requirements. To remove a module from the list, just remove its line. Python's `pip` command (which does the heavy lifting under the hood) will, of course, install dependencies as required. To keep things simple (this is a beginners' editor!) the notion of version number is ignored, and the latest version will be installed by default. Ergo, if a user needs to update a package, they should remove and re-install it.

Currently, there is no attempt to handle conflicts between pre-packaged (in the installer) modules or the same module installed via this method. My hunch is to remove these pre-installed modules in favour of this mechanism. **I would love to hear thoughts on this from folks!**

Packages installed in this way are available in both scripts run or debugged in Python 3 mode, or from within the Python 3 REPL (iPython shell).

# How Does it Work?

There were several ways to skin this cat, and I've gone for the simplest I can think of.

* The new tab for third party packages is added to the admin dialog. If, when the dialog is closed via "OK", there is a change in package state, then a new modal dialog launches to provide visibility of the package syncing process.
* This new package related dialog simply contains a text area into which is piped information from `pip` and other functionality.
* The **recommended way to use pip is by launching it in a subprocess**. The `pip` developers will not guarantee the internal API of the `pip` module. This is why things work the way they do: new packages are added to a list. The first new package is popped from the list and installed via `pip` in a subprocess. When the subprocess completes, if there are any more candidate packages to install, the process is repeated.
* The `--target` flag is used to specify an OS specific writeable directory into which these packages will be installed. This directory is added to the Python path when running or debugging Python 3 scripts or using the Python 3 REPL.
* Sadly, the `pip` command doesn't allow for targetting a "package directory" from which to remove packages. Ergo, Mu uses the list of packages that have been removed in the admin dialog to discover the metadata for them in the directory Mu uses for such packages. Happily, the metadata contains a list of all files related to each package. Mu simply iterates over these, deletes them and deletes any parent directories related to the module to be removed.

That's it! All feedback most welcome!

cc/@asweigart

ntoll committed:
First draft and tests for simple support for third party packages.

ntoll committed:
Ensure a reference to self.process

ntoll committed:
Fix layout issues.

ntoll committed:
Allow REPL access to third party modules installed via Mu.

bennuttall commented:
Just testing this on a Pi. It's a great idea and will be really helpful.

UI/UX: :+1: 

It seems to work nicely. I just added a couple of libraries and from the output, it seemed to do what I expected it to. It said it has succeeded.

However I tried to import a newly installed module and it failed (it seems the library was never installed). I went to try again and my list had disappeared. I'll investigate to see what went on, but that's all I know so far.

A few questions:

1. Where does it try to install new modules?
1. How do I upgrade a module I installed in Mu?
1. If pressing "OK" on the list of packages installs them, does it also upgrade new ones? (what if I want to / what if I don't want to?)
1. I assume these installations are at "user" level, i.e. local to the pi user, rather than as root with privilege escalation. Correct?

It might be worth pointing out that this is a pip installer, in case people either assume they can install apt packages, or they enter the apt package name "e.g. python3-abc" instead of "abc".

Really excited about this :1st_place_medal:

romilly commented:
I like the look of this *lots*. I'll have a play this week-end.

ntoll commented:
Where does it install packages..? Wherever this points to (from `logic.py`): 

`MODULE_DIR = os.path.join(DATA_DIR, 'site-packages')`.

On Linux that's likely to be: `~/.local/share/mu/site-packages/`

How to upgrade? Uninstall / reinstall.

No upgrades automagically (this really is super simple on purpose).

Yes... these are at user level (see path pasted above).

I've been testing it by poking around in the directory referenced above as I try to do different things. If there's an error, it should display it in package dialog's text area. If not, I've probably missed a setting to include `stderr` or something (likely). In any case, the log file may also contain information too. Under the hood it's just `pip` doing this: 

`python -m pip install foo --target /home/username/.local/share/mu/site-packages`

Where `python` is whatever Mu thinks is `sys.python`.

ntoll commented:
@bennuttall sorry, missed your apt related comment at the end. If they get confused by `apt` package names they're probably advanced enough to work it out for themselves. However, I agree mentioning "PyPI" will help mitigate problems.

ntoll committed:
Add PyPI clarification to UI instructions.

ntoll commented:
@bennuttall OK... I added the following line to the UI instructions: 

`Packages are installed from PyPI (see: https://pypi.org/).`

If that's not enough of a clue (along with link for further information) then they're in a right pickle. This will, of course, be documented and explained in the tutorials and howtos for 1.1.

ntoll committed:
Deal with different case module name / module directory and ensure all directories that don't contain any files are removed.

bennuttall commented:
If packages installed within Mu are only available within Mu, a user could find themselves confused when they try to launch their scripts from elsewhere, e.g. "I've finished my code, now let's have cron run it at boot".

What's the rationale for not installing into the (user's) library path?

carlosperate commented:
This looks great, thanks Nicholas!

I don't have much time to look into this PR properly today, I'll give it a proper go tomorrow or the day after, but I though I should give it a tough first test. So the first thing I tried was to install a module that Mu depends on and see what happens, in this case iPython.

Installing works fine, and doesn't seem to break the REPL, but encountered a couple of issues when uninstalling.
1. No matter which package from the list I remove, it always uninstalls and reinstalls Pygments as well (this happens to be the first item in the list, not sure if related or coincidence).
2. Removing all of them still leaves two packages present (ipython-genutils, prompt-toolkit) and I cannot remove them.
    * No errros given on user interface
    * Few errros shown in log:
    ```
    2019-01-30 00:12:47,612 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/pygmentize
    2019-01-30 00:12:47,612 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/pygmentize'
    2019-01-30 00:13:31,661 - mu.logic:1099(show_admin) INFO: Showing admin with logs from /Users/microbit-carlos/Library/Logs/mu/mu.log
    2019-01-30 00:13:36,899 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../share/man/man1/ipython.1.gz
    2019-01-30 00:13:36,899 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../share/man/man1/ipython.1.gz'
    2019-01-30 00:13:36,900 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/iptest
    2019-01-30 00:13:36,900 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/iptest'
    2019-01-30 00:13:36,900 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/iptest3
    2019-01-30 00:13:36,900 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/iptest3'
    2019-01-30 00:13:36,900 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/ipython
    2019-01-30 00:13:36,900 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/ipython'
    2019-01-30 00:13:36,900 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/ipython3
    2019-01-30 00:13:36,900 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/ipython3'
    2019-01-30 00:13:37,112 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/easy_install
    2019-01-30 00:13:37,112 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/easy_install'
    2019-01-30 00:13:37,112 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/easy_install-3.6
    2019-01-30 00:13:37,112 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/easy_install-3.6'
    2019-01-30 00:13:37,212 - mu.interface.dialogs:373(remove_package) ERROR: Unable to remove: /Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/pygmentize
2019-01-30 00:13:37,212 - mu.interface.dialogs:374(remove_package) ERROR: [Errno 2] No such file or directory: '/Users/microbit-carlos/Library/Application Support/mu/site-packages/../../bin/pygmentize'
```

GIF (heavy, 3 MBs) showing the test:
![2019-01-30 00 01 26](https://user-images.githubusercontent.com/4189262/51949144-4fa24280-2423-11e9-89a1-0afd3ad2471a.gif)

tjguk commented:
I've had the same result as @bennuttall: the usual pip output suggested that the install had worked. But "import networkzero as nw0" failed with an ImportError. I'll look into it later to see if I can see what's happening.

Win10, Python 3.6.4, pip 18.0, running from the checked-out branch

ntoll commented:
@bennuttall @tjguk hmm... I'll play around on my Windows laptop and see what's going on.

tjguk commented:
Well there's definitely something up with sys.path. This code (run under Mu):

```
import os, sys
for p in sys.path:
    print(p)
```

gives this result:

```
c:\users\goldent\mu_code
c:\users\goldent\mu_code\:C:\Users\goldent\.venvs\mu\Scripts\python36.zip:C:\Program Files\Python36\DLLs:C:\Program Files\Python36\lib:C:\Program Files\Python36:C:\Users\goldent\.venvs\mu:C:\Users\goldent\.venvs\mu\lib\site-packages:C:\Users\goldent\AppData\Local\python\mu\site-packages:C:\Users\goldent\.venvs\mu\lib\site-packages\IPython\extensions
C:\Users\goldent\.venvs\mu\Scripts\python36.zip
C:\Program Files\Python36\DLLs
C:\Program Files\Python36\lib
C:\Program Files\Python36
C:\Users\goldent\.venvs\mu
C:\Users\goldent\.venvs\mu\lib\site-packages
```

Note that second line, which is clearly a mis-munged path. (Looks like the colon-separator as wrongly been used to manipulate rather than the semicolon which is correct on Windows)

tjguk commented:
For completeness, the path where the packages should be and were installed is `C:\Users\goldent\AppData\Local\python\mu\site-packages` which is somewhere inside that jumbled second line (if you squint a bit)

ntoll commented:
@tjguk bingo. Cross platform inconsistencies FTW. I'll fix just as soon as I finish paying my tax bill! :-)

ntoll committed:
Fix PYTHONPATH generation from sys.path to use os.pathsep.

ntoll commented:
OK... so I have it working on Windows if I just `os.pathsep` instead of `:` for generating the `PYTHONPATH` from `sys.path`. I've also updated the unit tests to reflect this change. @tjguk can you please re-check..? Thank you.

ntoll commented:
@carlosperate aha... this is likely due to name mismatches between the advertised package name (found in the METADATA for the package) and the name used for identifying the module on the filesystem. I'll take a look.

ntoll committed:
Remove no-longer-needed Darwin specific PYTHONPATH munging.

ntoll commented:
@bennuttall but this is *still* beyond the scope of Mu. This is a learning opportunity for folks to understand that they need to install dependencies for their stuff. The simplest solution being to add the Mu `site-packages` to their system Python PATH. :-)

tjguk commented:
I think the crux here is that the internal pip functionality being added here is trying to solve what we might call a "walled-garden problem" for users of the bundled Mu distro who have the advantage of not having to install pre-installed packages but the disadvantage of not being able to install anything else!

Where a user is using a non-bundled Mu, they can install whatever they like in the usual way and it will be available to code run under Mu. Or am I wrong?

I mean: I do recognise the force of @bennuttall argument: that code which runs under Mu should not, ideally, fail to run outside Mu. I just don't know how much of an issue it would be. 

Thinking aloud about low-tech solutions: it wouldn't be the most complex thing in the world for Mu to run "pip freeze" and generate a requirements.txt file which could be run outside the Mu environment ("pip install --user -r requirements") to meet the need.

ntoll committed:
Fix to make correctly normalised package names for use with filesystem.

ntoll commented:
@carlosperate my hunch was correct, some of the package names were not getting normalised to the correct form for detecting their assets on the filesystem. E.g the package named `Foo-bar` has metadata contained in a directory called `foo_bar-1.0.0.dist-info`. I've updated the code and changed the relevant unit test to reflect this requirement. Please try again. Thank you.

carlosperate commented:
> Thinking aloud about low-tech solutions: it wouldn't be the most complex thing in the world for Mu to run "pip freeze" and generate a requirements.txt file which could be run outside the Mu environment ("pip install --user -r requirements") to meet the need.

This sounds like something that could go into the website, rather than a Mu feature. Goes well with the idea of "graduating" and showing a way for users to move their development environment. The problem would be discoverability, as things that are not straightforward to explain (or infer from the app) are usually hard to search for and hard find.

I do agree that the Raspberry Pi use case is a bit more confusing. For Windows and macOS, Mu runs its own packaged Python and I think it is clear to understand that it is its own "Python bubble".
Raspberry Pi on the other hand uses the system python3, so a user "pip3 installing", or "apt-get install python3-x" packages will have them available within Mu, so it is a bit more complex to understand why it works one way and not the other.

I woudn't be against Mu installing packages in the system python *only* on Raspberry Pi, as that makes that environment easier to understand, but that would require dealing with that specific use-case in the code and adds complexity to the codebase, it's not a lot, but these things do add up.

ntoll commented:
@tjguk yes, you are correct. If the editor was installed in a non-bundled way, whatever is available to system Python (or within your virtualenv) should just work.

Regarding @bennuttall's excellent point, this is *different* to how some of his own colleagues have discussed how Mu should work (isolated and self-contained with sensible defaults -- i.e. the completely understandable "don't make me think" problem for beginner developers who already have enough on their plate to think about). I believe there's enough of a simple solution (just install the packages you need via system Python's pip or via apt-get), that such a problem is easy to fix. The simplest one-time fix is to add Mu's `site-packages` to sys.path in one of several easy-to-do and well documented ways.

Does this make sense? I'm all ears for alternatives.

ntoll commented:
@carlosperate I think we're agreeing when I say this is a learning opportunity and you mention this is something for the website.

This feels **VERY** much like the "if you're asking about *that* sort of thing, you're probably too advanced for Mu" situation.  In which case we should try to help (docs on the website etc), but, as you put really well, this is an opportunity for "graduation" to the next level. Mu should encourage this sort of thing.

Finally, I'd feel very uncomfortable introducing yet more OS specific checks. We already do it for the plotter (because Raspbian's version of PyQt is from pre-history), and this just makes the code more complicated. Simple code is good code, and this sort of change doesn't sit well with me.

tjguk commented:
I think the key question is *how* to graduate in the sort of case which @bennuttall is describing. @ntoll is suggesting adding Mu's site-packages to sys.path in some way. This will (presumably) work but has three drawbacks that I see:

* I should be easy to work out where that is
* It won't work if you want to take a working project to a different machine
* It won't include any of Mu's own pre-installed modules (eg pygame etc.)

My suggestion was, in some way, to generate a requirements.txt. This, I think, overcomes the disadvantages I listed above, assuming that we have a simple way of achieving it from the new admin pane. @carlosperate thinks this should be docs rather than functionality. I'm fine with people learning -- in fact I'm proposing that they do indeed learn how a pip reqts file works etc. But the only way I can see for them to generate the file for themselves is to do something like this in Mu:

```
import subprocess
subprocess.run("pip freeze")
```

Obviously, that could be documented, but seems to me a bit clumsy (and it's not something I'd expect anyone to do in other circs) as opposed to us having an option, or always displaying the current list or something.

Thoughts?

carlosperate commented:
I also agree this is a learning opportunity. My main concern is that these learning opportunities usually end up in frustration.
No user is going to read the website from a to z, and they won't know how to figure out these things (a lot of users will not even know the website exists, they are just double clicking on a pre-installed application on the school or home computer). So what I think will end up happening is that we will have people not knowing why their code doesn't work outside of Mu.

Having this on the website is fine, but I think it is very important for this content to be discoverable in the app. That might end up being a link on the settings window, but that's an important link to have.

---

About the remove pacakges fix, thanks @ntoll!
Tested the latest commit and that removes the packages correctly.

I has playing with it installing `pyocd` and `ipython`, removed all the packages, and the settings window doesn't have any packages shown (great!).

However, looking at the site_packages folder, these files are still present:

<img width="959" alt="image" src="https://user-images.githubusercontent.com/4189262/51977133-7e99d200-247e-11e9-9e11-f022d7c9e6b6.png">

It's possible this is because to the packages metadata not listing all the files they install, in which case is not really for Mu to fix, but perhaps there is something to look into? In case there is a missing clean-up step that could be implemented.
If a setup.py file adds an entry point for a console_script, that will create those files in the bin directory, is that something Mu can remove as well? I would have thought pip would deal with these kind of files.

ntoll commented:
@tjguk hehehe... I love this sort of discussion. I **feels** like we're making progress (even if we may not be).

OK... how about we dump the equivalent of pip freeze into the logs..? That should just be "cut and pastable" and in a reasonably easy place to find.

ntoll commented:
@tjguk ack. :-) I'm going to take the dog for a walk and think on this.

holdenweb commented:
This may be a naiive suggestion, but when mu contains its own Python interpreter, would it be possible to provide a `mupython` command that ran the same interpreter in command-line mode? This seems to me like the easiest way to allow run-on-boot commands that use the same environment as mu.

tjguk commented:
@holdenweb it's a fine line, but I think that's moving Mu from being an editor into being an ecosystem, which I don't think is A Good Thing. [*]

Also, it's only really kicking the can down the road a bit: if someone wants to take that code and run it on a machine without Mu installed at all (or with Mu installed with a completely different set of additional packages) then you're *still* going to have to come up with a solution.

[*] Yes, that was a "1066 And All That" reference for those old enough to think it might be :)

bennuttall commented:
> but this is still beyond the scope of Mu.

I don't think it should be. Perhaps there's not much of a case for it on a traditional PC, but the Pi is a different beast.

Plus, my mentality for "zero" stuff is that I try to provide a smooth learning curve that helps them get started but let them keep going rather than get stuck at the edges of the library. there was a great blog posts about this concept recently: [Escape Hatches and Ejector Seats](https://anvil.works/blog/escape-hatches-and-ejector-seats). Sure, people will eventually move on from Mu, but I would hope they don't hit the edges too quick!

> This is a learning opportunity for folks to understand that they need to install dependencies for their stuff.

"But I already did"

> The simplest solution being to add the Mu site-packages to their system Python PATH. :-)

Not simple :)

> My main concern is that these learning opportunities usually end up in frustration.

Absolutely! Using IDLE is a learning opportunity...

> it's a fine line, but I think that's moving Mu from being an editor into being an ecosystem, which I don't think is A Good Thing.

Indeed. Simple is better than complex.

When you have multiple Python installations and various package dependencies to think about, you isolate projects into virtual environments With Raspberry Pi, the SD card *is* your isolated environment. There's a system python, and you generally install pip packages system-wide with sudo.

> Regarding @bennuttall's excellent point, this is different to how some of his own colleagues have discussed how Mu should work

I think in this context, my colleagues wouldn't want this side-effect.

I understand the rationale for Mu isolating its environment for Windows users, but the Debian/Raspbian version would use system python and should use system pip. On Raspbian, you know users have access to the terminal and pip. If Mu's package manager installed for the pi user, that would be fine, but if it only installs within Mu, it's just going to cause confusion.

Love the idea - don't want to throw a spanner in the works - just pointing out common use case from my experience!

ntoll commented:
*sigh* 

> Pi is a different beast

OK... in this case (just like with the plotter) perhaps the simplest thing to do is not show this admin option on the Pi. Folks can just `apt-get` their stuff as per standard Debian practices. Then the problem magically disappears.

ntoll commented:
@bennuttall on that note... what does `sys.info` return on a Pi so we can identify and make the relevant changes to the admin dialog..?

ntoll commented:
FWIW... when it comes to the plotter we set a flag on the result of importing the chart library from PyQt, so this approach won't work for detecting a Pi.

ntoll commented:
Apologies... I think I meant something like:

```import platform

p = platform.platform()```

And using the content of `p` to work out if I'm running in Raspbian... IYSWIM

bennuttall commented:
> OK... in this case (just like with the plotter) perhaps the simplest thing to do is not show this admin option on the Pi. Folks can just apt-get their stuff as per standard Debian practices. Then the problem magically disappears.

I think that's best - it's a great feature but I think more trouble than it's worth for Pi users. Just trying to protect people from getting a weird experience.

> @bennuttall on that note... what does ~sys.info~platform.platform() return on a Pi so we can identify and make the relevant changes to the admin dialog..?

```
Linux-4.14.79-v7+-armv7l-with-debian-9.6
```

Could be any kernel version, debian version or armv6 or armv7, so a match for "armv" should catch it.

> FWIW... when it comes to the plotter we set a flag on the result of importing the chart library from PyQt, so this approach won't work for detecting a Pi.

When Raspbian Buster arrives, we will magically get pyqtchart and the plotter should just work :+1:

ntoll committed:
Packaging is disabled on Raspberry Pi.

ntoll commented:
@bennuttall just pushed a fix for this. Can you please re-check?

Folks, I'd like to merge this into master so we can release an alpha 1 next week. Any reason NOT to do this? If I don't hear otherwise, I'll merge on Monday. :-)

carlosperate commented on mu/interface/dialogs.py:
```diff
@@ -224,7 +225,11 @@ def setup(self, log, settings, packages):
+            # Only allow third party package management if NOT running on a
+            # Raspberry Pi. See:
+            # https://github.com/mu-editor/mu/pull/749#issuecomment-459051823
+            self.tabs.addTab(self.package_widget, _('Third Party Packages'))
```
I think a missing tab in the settings depending on the platform could cause confusion for online material, tutorials or users trying to help other users.
I've often insisted to other people to go a specific setting in an app just to find out that the instructions I was given them were wrong because the feature was missing or somewhere else in their version they had.

Could we instead show the tab but change the content, a simple one liner saying something along of the lines of "In this platform packages can be installed in using pip3."? That way users would not wonder if they have an old version of Mu, and also will point them to the right direction.

ntoll replied:
Right... so we're just going to go round and round in circles about this. I believe we have a good-enough solution right now, and we should get this functionality in front of users as soon as possible as an alpha release so we get *feedback* from real people. Anything else is "educated" speculation (but speculation nonetheless). 

Only with feedback can we make an properly informed decision. If we don't get feedback, we're probably OK as we are. :-)

carlosperate replied:
I would say the Plotter not being available in Raspberry Pi is a good example of missing features in different platforms causing confusion:
https://github.com/mu-editor/mu/issues/714
https://github.com/mu-editor/mu/issues/689
https://github.com/mu-editor/mu/issues/694

tjguk commented:
I don't feel strongly about this, and I definitely support @ntoll approach of: let's get it out there. But, turning that back around again: why not "get it out there" with the same pip support enabled for RPi as for everyone else. We (via @bennuttall or elsewhere) can get feedback from real users who are or aren't confused and can decide on that basis.

Anyhow, having said my piece on that I'll leave @ntoll to decide.

carlosperate commented on mu/logic.py:
```diff
@@ -1074,28 +1096,51 @@ def show_admin(self, event=None):
+                    self.microbit_runtime = runtime
+                new_packages = [p for p in
+                                new_settings['packages'].lower().split('\n')
+                                if p.strip()]
```
If only the `new_packages` list is lower cased (and `packages` left with mixed case) you can end up with false positives when the two lists are compared in `self.sync_package_state()`.

ntoll replied:
The convention from Python packages means the result of `installed_packages` stored in `packages` will always contain lower-case package names.

carlosperate replied:
Pygments appears in that list with a capital letter, so if I guess the standard isn't enforced, or something else is going on.

ntoll replied:
So I've just tried installing and removing Pygmants. It works fine. 

I'm not sure what notion of a "false positive" you're getting at here. Looking at the code, the set operations and the way the directories ending in `dist-info` are used, I honestly don't believe this is a problem. 

Please can you give me a **concrete** example of where a failure occurs.

ntoll replied:
By concrete I mean repeatable steps I can follow to observe the false positive. Otherwise I'm investing lots of time in trying (and guessing) to recreate the "problem state", and that's no fun for anyone. :-(

carlosperate replied:
I've created this GIF (big 4Mbs, better viewed in its own tab), hope it makes things more clear, please do let me know if it doesn't and I'll provide more detailed steps.

![2019-02-12 15 35 19](https://user-images.githubusercontent.com/4189262/52647289-58435000-2edc-11e9-82cb-b590d813c847.gif)

In essence 'Pygments' is constantly uninstalled and reinstalled every time the settings window is closed because of the capitalisation. So the package can be installed and removed correctly, but every time the list is checked it will go through that cycle.

ntoll replied:
@carlosperate you :star: ... Now I **SEE** what you mean. Thank you. I'll fix this tomorrow (dad-duty beckons)..!

ntoll replied:
OK... fixed with latest push. Please re-test. Thank you! (Now to try to fix the damn builds)

carlosperate commented on mu/logic.py:
```diff
@@ -1074,28 +1096,51 @@ def show_admin(self, event=None):
-                                "Please try again.").format(runtime)
-                self._view.show_message(message, information)
+            new_settings = self._view.show_admin(logfile.read(), settings,
+                                                 '\n'.join(packages))
```
Don't think this is really a problem, but everything below this line could be unindented out of the `with` statement, so that the log file handler can be closed.
Line 1129 at the end of this function (inside this `with` statement) does also write to the log, but it seems to work fine.

ntoll replied:
You're right! Good spot. :-)

carlosperate commented:
I was testing the macOS packaged version and it crashed every time I tried to open the settings.

<details>
<summary>Log</summary>
<pre>Starting Mu 1.0.2
2019-02-01 20:51:20,169 - root:112(run) INFO: uname_result(system='Darwin', node='Carlos-MBP.local', release='18.2.0', version='Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64', machine='x86_64', processor='i386')
2019-02-01 20:51:20,169 - root:113(run) INFO: Python path: ['/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/MacOS', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/python/lib/python36.zip', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/python/lib/python3.6', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/python/lib/python3.6/lib-dynload', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/python/lib/python3.6/site-packages', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/app', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/app_packages', '/Users/microbit-carlos/Library/Application Support/mu/site-packages', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/app_packages/IPython/extensions']
2019-02-01 20:51:20,169 - root:114(run) INFO: Language code: en_GB
2019-02-01 20:51:20,383 - mu.logic:562(__init__) INFO: Setting up editor.
2019-02-01 20:51:20,383 - mu.logic:584(__init__) INFO: Settings path: /Users/microbit-carlos/Library/Application Support/mu/settings.json
2019-02-01 20:51:20,383 - mu.logic:585(__init__) INFO: Session path: /Users/microbit-carlos/Library/Application Support/mu/session.json
2019-02-01 20:51:20,383 - mu.logic:586(__init__) INFO: Log directory: /Users/microbit-carlos/Library/Logs/mu
2019-02-01 20:51:20,383 - mu.logic:587(__init__) INFO: Data directory: /Users/microbit-carlos/Library/Application Support/mu
2019-02-01 20:51:20,396 - mu.logic:600(setup) INFO: Available modes: python, adafruit, microbit, debugger, pygamezero
2019-02-01 20:51:20,509 - mu.logic:1195(change_mode) INFO: Workspace directory: /Users/microbit-carlos/mu_code
2019-02-01 20:51:20,515 - mu.logic:649(restore_session) INFO: Restoring session from: /Users/microbit-carlos/Library/Application Support/mu/session.json
2019-02-01 20:51:20,515 - mu.logic:650(restore_session) DEBUG: {'theme': 'day', 'mode': 'python', 'paths': ['/Users/microbit-carlos/mu_code/packages_test.py'], 'envars': [], 'minify': False, 'microbit_runtime': '', 'zoom_level': 2}
2019-02-01 20:51:20,515 - mu.logic:737(_load) INFO: Loading script from: /Users/microbit-carlos/mu_code/packages_test.py
2019-02-01 20:51:20,515 - mu.logic:293(read_and_decode) DEBUG: Trying to decode with utf-8
2019-02-01 20:51:20,515 - mu.logic:296(read_and_decode) INFO: Decoded with utf-8
2019-02-01 20:51:20,516 - mu.logic:309(read_and_decode) DEBUG: Detected newline '\n'
2019-02-01 20:51:20,516 - mu.logic:811(_load) DEBUG: # Write your code here :-)
print('hello world')
2019-02-01 20:51:20,578 - mu.logic:671(restore_session) INFO: Loaded files.
2019-02-01 20:51:20,578 - mu.logic:675(restore_session) INFO: User defined environment variables: []
2019-02-01 20:51:20,578 - mu.logic:679(restore_session) INFO: Minify scripts on micro:bit? False
2019-02-01 20:51:20,588 - mu.logic:1195(change_mode) INFO: Workspace directory: /Users/microbit-carlos/mu_code
2019-02-01 20:51:22,962 - mu.logic:1099(show_admin) INFO: Showing admin with logs from /Users/microbit-carlos/Library/Logs/mu/mu.log
2019-02-01 20:51:22,962 - root:94(excepthook) ERROR: Unrecoverable error
Traceback (most recent call last):
  File "/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/app/mu/logic.py", line 1107, in show_admin
    packages = installed_packages()
  File "/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/011FB62A-A14D-48E4-8D23-F9176A545338/d/mu-editor.app/Contents/Resources/app/mu/logic.py", line 160, in installed_packages
    lines = f.readlines()
  File "/Users/travis/build/mu-editor/mu_portable_python_macos/python/lib/python3.6/encodings/ascii.py", line 26, in decode
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 2616: ordinal not in range(128)
2019-02-01 20:51:43,773 - root:112(run) INFO: 
</pre>
</details>

<br>
The problem went away after I removed the 'site_packages' folder previously created by Mu running from this branch.

However I then (in the packaged Mu) installed a module, close the settings and had the same crash when trying to open the settings again in the same session.

<details>
<summary>Log</summary>
<pre>
2019-02-01 21:05:37,085 - root:111(run) INFO: 

-----------------

Starting Mu 1.0.2
2019-02-01 21:05:37,085 - root:112(run) INFO: uname_result(system='Darwin', node='Carlos-MBP.local', release='18.2.0', version='Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64', machine='x86_64', processor='i386')
2019-02-01 21:05:37,085 - root:113(run) INFO: Python path: ['/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/MacOS', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/python/lib/python36.zip', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/python/lib/python3.6', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/python/lib/python3.6/lib-dynload', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/python/lib/python3.6/site-packages', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app_packages', '/Users/microbit-carlos/Library/Application Support/mu/site-packages', '/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app_packages/IPython/extensions']
2019-02-01 21:05:37,085 - root:114(run) INFO: Language code: en_GB
2019-02-01 21:05:37,398 - mu.logic:562(__init__) INFO: Setting up editor.
2019-02-01 21:05:37,398 - mu.logic:582(__init__) DEBUG: Creating directory: /Users/microbit-carlos/Library/Application Support/mu/site-packages
2019-02-01 21:05:37,399 - mu.logic:335(get_admin_file_path) DEBUG: Creating admin file: /Users/microbit-carlos/Library/Application Support/mu/settings.json
2019-02-01 21:05:37,399 - mu.logic:584(__init__) INFO: Settings path: /Users/microbit-carlos/Library/Application Support/mu/settings.json
2019-02-01 21:05:37,399 - mu.logic:335(get_admin_file_path) DEBUG: Creating admin file: /Users/microbit-carlos/Library/Application Support/mu/session.json
2019-02-01 21:05:37,399 - mu.logic:585(__init__) INFO: Session path: /Users/microbit-carlos/Library/Application Support/mu/session.json
2019-02-01 21:05:37,399 - mu.logic:586(__init__) INFO: Log directory: /Users/microbit-carlos/Library/Logs/mu
2019-02-01 21:05:37,399 - mu.logic:587(__init__) INFO: Data directory: /Users/microbit-carlos/Library/Application Support/mu
2019-02-01 21:05:37,410 - mu.logic:600(setup) INFO: Available modes: python, adafruit, microbit, debugger, pygamezero
2019-02-01 21:05:37,544 - mu.logic:1195(change_mode) INFO: Workspace directory: /Users/microbit-carlos/mu_code
2019-02-01 21:05:37,548 - mu.logic:649(restore_session) INFO: Restoring session from: /Users/microbit-carlos/Library/Application Support/mu/session.json
2019-02-01 21:05:37,548 - mu.logic:650(restore_session) DEBUG: {}
2019-02-01 21:05:37,549 - mu.logic:1152(select_mode) INFO: Showing available modes: ['python', 'adafruit', 'microbit', 'debugger', 'pygamezero']
2019-02-01 21:05:38,897 - mu.logic:702(restore_session) INFO: Starting with blank file.
2019-02-01 21:05:39,010 - mu.logic:1195(change_mode) INFO: Workspace directory: /Users/microbit-carlos/mu_code
2019-02-01 21:05:41,167 - mu.logic:1099(show_admin) INFO: Showing admin with logs from /Users/microbit-carlos/Library/Logs/mu/mu.log
2019-02-01 21:06:02,062 - mu.logic:1099(show_admin) INFO: Showing admin with logs from /Users/microbit-carlos/Library/Logs/mu/mu.log
2019-02-01 21:06:02,064 - root:94(excepthook) ERROR: Unrecoverable error
Traceback (most recent call last):
  File "/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app/mu/logic.py", line 1107, in show_admin
    packages = installed_packages()
  File "/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app/mu/logic.py", line 160, in installed_packages
    lines = f.readlines()
  File "/Users/travis/build/mu-editor/mu_portable_python_macos/python/lib/python3.6/encodings/ascii.py", line 26, in decode
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 2616: ordinal not in range(128)
</pre>
</details>

```
2019-02-01 21:06:02,064 - root:94(excepthook) ERROR: Unrecoverable error
Traceback (most recent call last):
  File "/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app/mu/logic.py", line 1107, in show_admin
    packages = installed_packages()
  File "/private/var/folders/l1/szsxrhfj7v1fglxfv9b35lz80000gn/T/AppTranslocation/A6CF3417-08AD-48D7-87EC-60E57CF4CCD0/d/mu-editor.app/Contents/Resources/app/mu/logic.py", line 160, in installed_packages
    lines = f.readlines()
  File "/Users/travis/build/mu-editor/mu_portable_python_macos/python/lib/python3.6/encodings/ascii.py", line 26, in decode
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 2616: ordinal not in range(128)
```

I've also noticed the log doesn't include any entries during the installation process, would be useful to have something logged for debugging.

tjguk commented:
Thanks for building, @carlosperate . I've just tested the Win64 build and it seems to work ok. (Added a package; removed a package).

bennuttall commented:
I had a quick look over the code earlier and noticed you delete package files based on the package name as installed - you can't assume that package files will match the package name at all unfortunately - for example the package `opencv-python` is `cv2`. How come you're not using pip to uninstall?

tjguk commented:
@bennuttall I think that's explained in the final bullet point under "How does it work?" above.

ntoll commented:
@carlosperate awesome commentary there matey... I'll try to address each of your points by the end of the weekend. 

WRT confused users... I'm at a loss for what to do. We **clearly state** in our docs on the website that the plotter doesn't work on Raspberry Pi. We have no way to change this situation since it's a problem to do with Raspbian's ancient PyQt package not having QtCharts. I guess your suggestion of keeping the tab but putting in a helpful "Feature not available on Raspberry Pi" message may save us some bug reports (but I also suspect we'll get bug reports that say, "make this tab work on Raspberry Pi"). ;-) So we're in a sort of no-win situation -- hence my desire to get this in front of real teachers and learners so they have a plenty of time to comment and provide feedback.

@bennuttall the `pip` command doesn't do targetted uninstall (see "How does it work"). :-( The files which are deleted are *only* those referenced from package metadata (i.e. Mu isn't guessing which files to delete). In terms of package name discovery, the *metadata* is always in a directory which matches a particular known pattern. So, for example, the metadata for `opencv-python` is in `opencv_python-4.0.0.21.dist-info` (I just checked to confirm this). The metadata within that directory tells Mu that the files to uninstall are in the `cv2` directory. I can confirm that Mu correctly installs and uninstalls `opencv-python` using this process of depending on metadata. Does this make sense? Put simply, deletion **isn't** done via pattern matching on the name (that would be dangerous!). ;-)

Here's what I'll  try to do next:

* Keep the tab in RPi, but with a "Not available" message instead of the functioning UI.
* Update, address, check @carlosperate's further code related comments.
* Wait for confirmation that the crash is fixed on OSX (it works for me on my Mac - I suspect on ascii characters in your path as the cause).
* Merge this (and some other outstanding PRs) with a view to Mu1.1 alpha 1 sometime next week.

As always, comments, constructive critique and ideas most welcome.

carlosperate commented:
I think the problem with the macOS packaged version crashing is to do with the content of the METADA file rather than the file path (fails on line 160):
https://github.com/mu-editor/mu/blob/aa346178f3e122e1a22e08822210c897fb1f152c/mu/logic.py#L157-L162

It can be replicated by installing any module, then editing its METADATA file to include a Chinese character (我) in the description area, and then trying to open the settings window.

ntoll commented:
@carlosperate OK... so what is in your problematic METADATA file (and how did it get there?). Could you attach it to this issue please? :-)

carlosperate commented:
Problem appeared for one of the ipython dependencies, which turned out to be `pexpect`. Metadata file attached: [METADATA.zip](https://github.com/mu-editor/mu/files/2825579/METADATA.zip)

ntoll commented:
I have a five hour train journey today (and tomorrow0 so will try to fix the remaining issues while speeding to and from Scotland. ;-)

ntoll committed:
Fix message in 3rd party tab for RPi users. Fix unicode problem from METADATA.

ntoll commented:
I've finally found the time to complete the work on this branch as I described.

The "Third Party Packages" tab appears in Raspberry Pi, but contains a "does not work on RPi" type message instead of the usual form. @bennuttall can you please check this works OK..?

@carlosperate I was **unable** to recreate your encoding problem, even when I used the faulty METADATA file you attached to this issue. I have no idea what's going on there. However, to mitigate this sort of problem, I've ensured the file is read with the `"rb"` flag (as bytes) and only then extract the package name from the first line. Please can you test this on your system. It *feels* like it could be something odd about your setup since it works *fine* for me on OSX. :-/

Once the CI is green I'll merge this as promised with a view to an imminent Alpha 1

carlosperate commented:
@ntoll were you using the final app bundle after building locally?
I only encounter this error with the packaged output, not when running with my local Python from the terminal.

ntoll commented:
@carlosperate aha... I see what you mean. Nope. As far as I can tell, my changes *should* fix it though. I guess there's some encoding being set in your terminal vs encoding used when launching an app. *sigh* I hate computers.

ntoll committed:
Fix failing tests due to Qt5.12 update.

ntoll commented:
*sigh* the update to Qt5.12 has broken the build. I have a meeting in 2 mins, but will try to fix later. :-/

bennuttall commented:
Sorry for the delay @ntoll - I can confirm that the third party package tab shows the "not supported" message on Raspberry Pi. Thanks again for dealing with the edge case :+1:

carlosperate commented:
> @carlosperate aha... I see what you mean. Nope. As far as I can tell, my changes _should_ fix it though. I guess there's some encoding being set in your terminal vs encoding used when launching an app.

Yeah, maybe an environmental variable somehow indicating default encoding, similar to the issue we had with detecting the default system locale.

Anyway, I've just built a mu-editor.app package locally and cannot replicate the MANIFEST decoding issue any more 😃 🎉

ntoll commented:
@carlosperate good stuff. I'll fix the false-positive stuff in the morning.

ntoll committed:
Fix false positives due to case sensitivity.

ntoll committed:
Travis fix

ntoll committed:
More Travis mess.

ntoll committed:
Fix missing package needed for Qt5.12 on Linux.

ntoll committed:
Argh typo

ntoll committed:
More Travis config changes...

ntoll committed:
Bump OSX version number to the next one supported by Apple/Travis.

ntoll committed:
OSX pyenv install->upgrade.

ntoll commented:
OK... I finally got the Travis related problems fixed so will merge back to master now that the outstanding issues have been addressed.

Thank you to everyone for all your help with the creation of this feature. I really really appreciate the your time, care and attention taken to improve this PR from my shonky bodge. ;-)

Onwards and upwards...

Daniels118 commented:
On Windows, if adding packages fails with error `CERTIFICATE_VERIFY_FAILED`, then you may append base64-encoded custom certificates to the file
`%LocalAppData%\python\mu\mu_venv-xx-xxxxxxxx-xxxxxx\Lib\site-packages\pip\_vendor\certifi\cacert.pem`.
